### DISCUSSION TITLE
socket.handshake.headers vs socket.request.headers

### DISCUSSION THREAD

**Discussion by thernstig** [socketio/socket.io#4607] [Q&A] [answered]
I cannot from https://socket.io/docs/v4/server-api/#socketrequest and https://socket.io/docs/v4/server-api/#sockethandshake deduct what the differences are.

Is `socket.request.headers` only ever set once?
And socket.handshake.headers also only set once in case the transport uses Websocket, but set each time on the HTTP request if using HTTP long polling?

**Comment by darrachequesne** [comment, marked as answer]
Hi!

`socket.handshake.headers` and `socket.request.headers` are the exact same object.

https://github.com/socketio/socket.io/blob/0c0eb0016317218c2be3641e706cfaa9bea39a2d/lib/socket.ts#L288-L290

> Is socket.request.headers only ever set once?

Yes, those are the headers of the first HTTP request.

> And socket.handshake.headers also only set once in case the transport uses Websocket, but set each time on the HTTP request if using HTTP long polling?

No, they are set once in all cases.

**Reply by thernstig** [reply]
Thanks for the reply!

If they are the same object, how would one get any new headers sent when using HTTP long polling? Imagine someone injects headers via a proxy for each HTTP long-polling request and they could differ.

Could the docs at https://socket.io/docs/v4/server-api/#socketrequest and https://socket.io/docs/v4/server-api/#sockethandshake be updated that this is the same object? Reading the descriptions I would have thought `socket.request.headers` actually could change and be updated during each HTTP long polling request.

**Reply by darrachequesne** [reply]
> Could the docs at https://socket.io/docs/v4/server-api/#socketrequest and https://socket.io/docs/v4/server-api/#sockethandshake be updated that this is the same object?

Done!

> If they are the same object, how would one get any new headers sent when using HTTP long polling?

There are no official way to do this, but you should be able to do something like this:

```js
io.on("connection", (socket) => {
  const transport = socket.conn.transport;
  const onRequest = transport.onRequest;

  transport.onRequest = (req) => {
    const headers = req.headers;
    // do something

    onRequest.call(transport, req);
  }
});
```
